db/keyval/filedb/filesystem/filesystem.go

```diff
@@ -0,0 +1,180 @@
+	GetFileNames(paths []string) ([]string, error)
+	// Watch given paths, call 'onEvent' function on every event or 'onClose' if watcher is closed
+	Watch(paths []string, onEvent func(event fsnotify.Event), onClose func()) error
+	// Close allows to terminate watcher from outside which calls 'onClose'
+	Close() error
+}
+
+// FsHandler is helper struct to manipulate with filesystem API
+type FsHandler struct {
+	log     logging.Logger
+	watcher *fsnotify.Watcher
+}
+
+// NewFsHandler creates a new instance of file system handler
+func NewFsHandler() *FsHandler {
+	return &FsHandler{}
+}
+
+// CreateFile is an implementation of the file system API interface
+func (fsh *FsHandler) CreateFile(file string) error {
+	parts := strings.Split(file, "/")
+	if len(parts) == 0 {
+		return errors.Errorf("failed to create file: invalid input %s", file)
+	}
+	// Create path at first if necessary
+	if len(parts) > 1 {
+		path := strings.Replace(file, parts[len(parts)-1], "", 1)
+		if err := os.MkdirAll(path, os.ModePerm); err != nil {
+			return errors.Errorf("failed to create path for file %s: %v", file, err)
+		}
+	}
+	sf, err := os.Create(file)
+	if err != nil {
+		return errors.Errorf("failed to create file %s: %v", file, err)
+	}
+	return sf.Close()
+}
+
+// ReadFile is an implementation of the file system API interface
+func (fsh *FsHandler) ReadFile(file string) ([]byte, error) {
+	return ioutil.ReadFile(file)
+}
+
+// WriteFile is an implementation of the file system API interface
+func (fsh *FsHandler) WriteFile(file string, data []byte) error {
+	fileObj, err := os.OpenFile(file, os.O_TRUNC|os.O_WRONLY, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("failed to open status file %s for writing: %v", file, err)
+	}
+	defer fileObj.Close()
+	if _, err := fileObj.Write(data); err != nil {
+		return fmt.Errorf("failed to write status file %s for writing: %v", file, err)
+	}
+	return nil
+}
+
+// FileExists is an implementation of the file system API interface
+func (fsh *FsHandler) FileExists(file string) bool {
+	if _, err := os.Stat(file); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+// GetFileNames is an implementation of the file system API interface
+func (fsh *FsHandler) GetFileNames(paths []string) (files []string, err error) {
+	for _, path := range paths {
+		err := filepath.Walk(path, func(path string, info os.FileInfo, err error) error {
+			if info == nil || info.IsDir() {
+				return nil
+			}
+			files = append(files, path)
+			return nil
+		})
+		if err != nil {
+			err = errors.Errorf("failed to traverse through %s: %v", path, err)
+		}
+	}
+	return files, err
+}
+
+// Watch starts new filesystem notification watcher. All events from of json/yaml type files are passed to 'onEvent' function.
+// Function 'onClose' is called when event channel is closed.
+func (fsh *FsHandler) Watch(paths []string, onEvent func(event fsnotify.Event), onClose func()) error {
```

Comment:
Instead needing two callbacks, you could use single channel as parameter in which the events would be sent and could be closed on Close. Also it could also be buffered if needed.